ContinuousRegistration/Source/make_hyperopt.py

```diff
@@ -0,0 +1,220 @@
+from ContinuousRegistration.Source.make_registration_scripts import parser, run as make_registration_scripts
+from ContinuousRegistration.Source.make_evaluation import run as make_evaluation
+from ContinuousRegistration.Source.util import write_json
+
+from hyperopt import fmin, tpe, hp
+from ContinuousRegistration.Source.datasets import load_datasets
+import shutil, numpy as np
+
+import sys, os, time
+from subprocess import Popen, check_output
+import glob, copy
+
+def exec_commands(parameters, cmds):
+    ''' Exec commands in parallel in multiple process
+    (as much as we have CPU)
+    '''
+    if not cmds: return # empty list
+
+    def done(p):
+        return p.poll() is not None
+    def success(p):
+        return p.returncode == 0
+    def fail():
+        sys.exit(1)
+
+    processes = []
+    while True:
+        while cmds and len(processes) < parameters.jobs:
+            task = cmds.pop()
+            print('Running %s ...' % task)
+            processes.append(Popen(task))
+
+        for p in processes:
+            if done(p):
+                if success(p):
+                    processes.remove(p)
+                else:
+                    fail()
+
+        if not processes and not cmds:
+            break
+        else:
+            time.sleep(1.0)
+
+class objective():
+    def __init__(self, parameters):
+        self.parameters = parameters
+
+        # Preprocess data
+        self.datasets = load_datasets(self.parameters)
+
+    def __call__(self, hyperopts):
+        # Write new blueprint to disk
+        blueprint = copy.deepcopy(parameters.blueprint)
+
+        for key, value in hyperopts.items():
+            blueprint["Components"].append(value)
+
+        write_json(os.path.join(parameters.submissions_directory, parameters.team_name, self.parameters.blueprint_file_name[0]), blueprint)
+
+        # Make shell scripts for running registrations. Scripts from previous iterations will be overridden
+        make_registration_scripts(self.parameters)
+
+        # Run registrations
+        if sys.platform == 'win32':
+            try:
+                exec_commands(parameters,
+                              glob.glob(os.path.join(parameters.output_directory, '*', '*', 'bat', '*.bat')))
+            except Exception as e:
+                print(e)
+        else:
+            try:
+                exec_commands(parameters,
+                              glob.glob(os.path.join(parameters.output_directory, '*', '*', 'sh', '*.sh')))
+            except Exception as e:
+                print(e)
+
+        # Evaluate and report target metric
+        results = make_evaluation(parameters)
+        result = []
+        for team_results in results:
+            for blueprint_results in team_results:
+                for dataset_results in blueprint_results:
+                    for registration_results in dataset_results:
+                        result.append(registration_results[parameters.metric])
```

Comment:
what about rather using `itertools.chain`
https://docs.python.org/2/library/itertools.html#itertools.chain

Reply:
ok, another option is here: https://stackoverflow.com/a/38722308/4521646
they code flatten function and probably you would find it application also elsewhere :)